sdk/include/opentelemetry/sdk/trace/tracer.h

```diff
@@ -50,6 +50,9 @@ class Tracer final : public trace_api::Tracer, public std::enable_shared_from_th
     return *instrumentation_library_;
   }
 
+  /** Returns the currently configured resource **/
+  const opentelemetry::sdk::resource::Resource &GetResource() { return context_->GetResource(); }
```

Comment:
No setter - `SetResource` in API, only getter - `GetResource` ? What are the thread-safety guarantees of this method on API surface?

Reply:
The resource is a property of `TracerProvider` and should be set during the initialization of TracerProvider in the main thread. It's not re-settable again once initialized.

Reply:
I approved.

From a practical standpoint, however, we do have scenarios as follows:
- whatever Monitoring Agent is upgraded, OR remote configuration is upgraded.
- your `resource` contents now may need to change - to indicate that a new configuration is applied without application service restart.

In those cases, I think, we may need to consider how to make some of this thread-safe. Current logic of returning `Resource` by reference in SDK may break if `Resource` needs to change in a different thread.

I am not sure how to track my concern. Maybe we need to have a generic umbrella issue that captures thread-safety guarantees at both levels:
- API thread safety guarantees
- SDK thread safety guarantees

In a separate markdown document, dev-focused document for exporter developers.

Reply:
> * your `resource` contents now may need to change, e.g. updated configuration of the `TracerProvider` - to indicate that a new configuration is now applied. Without application service restart. Your threads continue running, while accessing resource, while the resource object is maybe changing at the same time.. which may lead to crash.

The ownership model of the Resource object is well defined. Once created, the ownership of the Resource object is transferred to TracerProvider, and there is no setters to update it at runtime.

Reply:
Also to add, as per the specs : https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/sdk.md

``
When used with distributed tracing, a resource can be associated with the TracerProvider when the TracerProvider is created. That association cannot be changed later. When associated with a TracerProvider, all Spans produced by any Tracer from the provider MUST be associated with this Resource.
``

The resource once set can't be changed later.